File: src/familiar_connect/focus.py
"""Attentional focus controller for a single familiar."""

from __future__ import annotations

import asyncio
import logging
import time
from typing import TYPE_CHECKING

from familiar_connect import log_style as ls
from familiar_connect.subscriptions import SubscriptionKind

if TYPE_CHECKING:
    from collections.abc import Awaitable, Callable

    from familiar_connect.history.async_store import AsyncHistoryStore
    from familiar_connect.subscriptions import SubscriptionRegistry

_logger = logging.getLogger(__name__)

# Debounce window: rapid arrivals within this window share one nudge
_DEFAULT_NUDGE_DEBOUNCE_S = 30.0


class FocusManager:
    """Per-familiar attentional focus controller.

    Two independent focus pointers (text, voice). Focus shifts are
    model-decided (via shift_focus tool) and applied immediately at
    tool-call time under per-modality lock — no deferral. A silent turn
    therefore still leaves her where she went; there is no pending
    state to leak into a later turn.

    Unread nudge: when a non-focused channel gets traffic, ``should_wake``
    flags that the model deserves a turn (the responder fires a synthetic
    wake) — the arrival itself fires it, with no idle-silence requirement.
    ``unread_nudge_enabled`` gates the behavior on/off; the debounce window
    is the sole throttle. The nudge never moves focus — only the model's
    shift_focus does. Rapid arrivals within ``nudge_debounce_seconds`` are
    grouped into one nudge; the next unread after the window fires again.
    """

    def __init__(
        self,
        *,
        familiar_id: str,
        store: AsyncHistoryStore,
        subscriptions: SubscriptionRegistry,
        clock: Callable[[], float] = time.monotonic,
        unread_nudge_enabled: bool = True,
        nudge_debounce_seconds: float = _DEFAULT_NUDGE_DEBOUNCE_S,
    ) -> None:
        self._familiar_id = familiar_id
        self._store = store
        self._subscriptions = subscriptions
        self._text_focus: int | None = None
        self._voice_focus: int | None = None
        self._text_lock = asyncio.Lock()
        self._voice_lock = asyncio.Lock()
        # Unread-nudge state
        self._clock = clock
        self._unread_nudge_enabled = unread_nudge_enabled
        self._nudge_debounce_seconds = nudge_debounce_seconds
        self._last_nudge: float = float("-inf")  # Never nudged initially
        # Channel_id → display name; populated by bot on_ready
        self.channel_names: dict[int, str] = {}
        # Channel_id → guild name; populated by bot on_ready
        self.guild_names: dict[int, str] = {}
        # Called once after each applied shift; None disables
        self.on_shift: Callable[[], Awaitable[None]] | None = None

    async def initialize(self) -> None:
        """Load persisted focus pointers from DB.

        Pointers to channels no longer subscribed are dropped (avoid
        stranding focus on a dead channel); startup fallback re-seeds.
        """
        ptrs = await self._store.get_focus_pointers(self._familiar_id)
        if ptrs is not None:
            self._text_focus = self._keep_if_subscribed(ptrs.text_channel_id)
            self._voice_focus = self._keep_if_subscribed(ptrs.voice_channel_id)
            _logger.info(
                f"{ls.tag('Focus', ls.LC)} loaded "
                f"{ls.kv('text', self.channel_label(self._text_focus), vc=ls.LW)} "
                f"{ls.kv('voice', self.channel_label(self._voice_focus), vc=ls.LW)}"
            )

    def _keep_if_subscribed(self, channel_id: int | None) -> int | None:
        """Pass through channel_id when subscribed; else None (warn on drop)."""
        if channel_id is None or self.is_subscribed(channel_id):
            return channel_id
        _logger.warning(
            f"{ls.tag('Focus', ls.LC)} dropped stale pointer "
            f"{ls.kv('channel', self.channel_label(channel_id), vc=ls.LW)} "
            "(no longer subscribed)"
        )
        return None

    def get_focus(self, modality: str) -> int | None:
        """Return current focus channel_id for modality."""
        return self._text_focus if modality == "text" else self._voice_focus

    def is_subscribed(self, channel_id: int) -> bool:
        """Whether channel_id is a known text/voice subscription."""
        return self._subscriptions.kind_for(channel_id) is not None

    def subscribed_channels(self) -> list[int]:
        """Sorted, deduped channel_ids across text + voice subscriptions."""
        return sorted({sub.channel_id for sub in self._subscriptions.all()})

    def is_focused(self, channel_id: int) -> bool:
        """Check if channel_id is the active text or voice focus."""
        return channel_id in {self._text_focus, self._voice_focus}

    async def shift_now(self, channel_id: int) -> None:
        """Apply focus shift immediately (at tool-call time).

        Modality inferred from subscriptions. Text shift promotes the
        target's staged backlog to consumed, moves the pointer, persists,
        and fires ``on_shift``. No deferral — focus moves now, so a silent
        turn still leaves her where she went and nothing leaks.
        """
        kind = self._subscriptions.kind_for(channel_id)
        modality = "voice" if kind is SubscriptionKind.voice else "text"
        lock = self._voice_lock if modality == "voice" else self._text_lock
        async with lock:
            if modality == "text":
                count = await self._store.promote_staged_turns(
                    familiar_id=self._familiar_id, channel_id=channel_id
                )
                self._text_focus = channel_id
                _logger.info(
                    f"{ls.tag('🔀 Focus', ls.LC)} "
                    f"{ls.kv('text', self.channel_label(channel_id), vc=ls.LW)} "
                    f"{ls.kv('promoted', str(count), vc=ls.LG)}"
                )
            else:
                self._voice_focus = channel_id
                _logger.info(
                    f"{ls.tag('🔀 Focus', ls.LC)} "
                    f"{ls.kv('voice', self.channel_label(channel_id), vc=ls.LW)}"
                )
            await self._store.set_focus_pointers(
                self._familiar_id,
                text_channel_id=self._text_focus,
                voice_channel_id=self._voice_focus,
            )
        if self.on_shift is not None:
            await self.on_shift()

    def should_wake(self, channel_id: int) -> bool:
        """Whether a non-focused arrival warrants a nudge.

        True when nudges are enabled, the channel is unfocused, and we
        are outside the debounce window — the arrival itself fires the
        nudge, regardless of how recently the focused channel was active.
        Debounce is the sole throttle: rapid arrivals share one nudge.
        """
        if not self._unread_nudge_enabled:
            return False
        if self.is_focused(channel_id):
            return False
        now = self._clock()
        if (now - self._last_nudge) < self._nudge_debounce_seconds:
            return False
        return True

    def mark_nudge_pending(self) -> None:
        """Record nudge timestamp to start debounce window."""
        self._last_nudge = self._clock()

    async def end_turn(self) -> None:
        """Responder end-of-turn hook — intentionally a no-op.

        Focus shifts apply immediately (``shift_now``), so there is no
        per-turn focus state to reset. Retained because both responders
        call it as their end-of-turn signal; stays async so they can
        await it uniformly.
        """
        return

    def channel_label(self, channel_id: int | None) -> str:
        """Format channel_id as '#name(id)' or '#id' when name unknown."""
        if channel_id is None:
            return "none"
        name = self.channel_names.get(channel_id)
        return f"#{name}({channel_id})" if name else f"#{channel_id}"

    def guild_name_for(self, channel_id: int | None) -> str | None:
        """Server name for channel_id; None for None input or unknown channel."""
        return self.guild_names.get(channel_id) if channel_id is not None else None

    def presence_guild(self) -> str | None:
        """Guild name for current text focus channel; None when unset or unknown."""
        return self.guild_name_for(self._text_focus)

    def presence_text(self) -> str | None:
        """'#channel-name' for current text focus; None when unset."""
        channel_id = self._text_focus
        if channel_id is None:
            return None
        name = self.channel_names.get(channel_id, str(channel_id))
        return f"#{name}"

    def set_focus_immediately(self, channel_id: int, modality: str) -> None:
        """Set focus without deferral (used at startup)."""
        if modality == "text":
            self._text_focus = channel_id
        else:
            self._voice_focus = channel_id
        _logger.info(
            f"{ls.tag('Focus', ls.LC)} default "
            f"{ls.kv(modality, self.channel_label(channel_id), vc=ls.LW)}"
        )

File: familiar-connect/src/focus.rs
//! Attentional focus controller for a single familiar (subsystem 05; Python
//! `focus.py`).
//!
//! Two independent focus pointers (text, voice). Focus shifts are model-decided
//! (via the `shift_focus` tool) and applied immediately at tool-call time — no
//! deferral (behavior 50; `end_turn` is a no-op). An unread nudge fires when a
//! non-focused channel receives traffic, throttled by a debounce window with an
//! injectable clock (DESIGN §4.8).
//!
//! Per DESIGN port notes the two Python `asyncio.Lock`s collapse into a single
//! [`std::sync::Mutex`] over the focus state — an acceptable simplification that
//! also closes the benign cross-modal double-pointer persist race (behavior 49).
//! `on_shift` is invoked outside the lock. The store dependency is a narrow
//! [`FocusStore`] trait so tests inject a scripted double.

use std::collections::{BTreeSet, HashMap};
use std::sync::{Arc, Mutex};
use std::time::Instant;

use async_trait::async_trait;
use futures::future::BoxFuture;

use crate::history::StoreError;
use crate::history::async_store::AsyncHistoryStore;
use crate::history::store::{FocusPointers, Promotion};
use crate::log_style as ls;
use crate::subscriptions::{SubscriptionKind, SubscriptionView};

/// Default debounce window: rapid arrivals within it share one nudge.
const DEFAULT_NUDGE_DEBOUNCE_S: f64 = 30.0;

/// Default catch-up window — staged turns taken in on a shift.
const DEFAULT_CATCH_UP_LIMIT: usize = 20;

/// Server label for DM channels admitted via the allowlist. This module owns
/// `guild_names` / `guild_name_for`, so the constant lives here.
pub const PRIVATE_MESSAGE_GUILD_NAME: &str = "Private Message";

/// Injectable monotonic clock returning seconds.
pub type Clock = Arc<dyn Fn() -> f64 + Send + Sync>;

/// Presence-refresh hook awaited once after each applied shift.
pub type OnShift = Arc<dyn Fn() -> BoxFuture<'static, ()> + Send + Sync>;

/// The narrow store seam [`FocusManager`] needs (DESIGN §4.8: tests inject a
/// scripted double). Implemented for [`AsyncHistoryStore`].
#[async_trait]
pub trait FocusStore: Send + Sync {
    /// Load the persisted focus pointers row, if any.
    async fn get_focus_pointers(
        &self,
        familiar_id: &str,
    ) -> Result<Option<FocusPointers>, StoreError>;
    /// Upsert both pointers.
    async fn set_focus_pointers(
        &self,
        familiar_id: &str,
        text_channel_id: Option<i64>,
        voice_channel_id: Option<i64>,
    ) -> Result<(), StoreError>;
    /// Promote a text channel's staged backlog to consumed on shift.
    async fn promote_staged_turns(
        &self,
        familiar_id: &str,
        channel_id: i64,
        catch_up_limit: usize,
    ) -> Result<Promotion, StoreError>;
}

#[async_trait]
impl FocusStore for AsyncHistoryStore {
    async fn get_focus_pointers(
        &self,
        familiar_id: &str,
    ) -> Result<Option<FocusPointers>, StoreError> {
        Self::get_focus_pointers(self, familiar_id.to_owned()).await
    }

    async fn set_focus_pointers(
        &self,
        familiar_id: &str,
        text_channel_id: Option<i64>,
        voice_channel_id: Option<i64>,
    ) -> Result<(), StoreError> {
        Self::set_focus_pointers(
            self,
            familiar_id.to_owned(),
            text_channel_id,
            voice_channel_id,
        )
        .await
    }

    async fn promote_staged_turns(
        &self,
        familiar_id: &str,
        channel_id: i64,
        catch_up_limit: usize,
    ) -> Result<Promotion, StoreError> {
        Self::promote_staged_turns(
            self,
            familiar_id.to_owned(),
            channel_id,
            Some(catch_up_limit),
        )
        .await
    }
}

/// Interior mutable focus state (single-mutex design; DESIGN port notes).
struct FocusState {
    text_focus: Option<i64>,
    voice_focus: Option<i64>,
    channel_names: HashMap<i64, String>,
    guild_names: HashMap<i64, String>,
    last_nudge: f64,
}

/// Per-familiar attentional focus controller.
pub struct FocusManager {
    familiar_id: String,
    store: Arc<dyn FocusStore>,
    subscriptions: Arc<dyn SubscriptionView>,
    clock: Clock,
    unread_nudge_enabled: bool,
    nudge_debounce_seconds: f64,
    catch_up_limit: usize,
    state: Mutex<FocusState>,
    on_shift: Mutex<Option<OnShift>>,
}

fn monotonic_clock() -> Clock {
    let start = Instant::now();
    Arc::new(move || start.elapsed().as_secs_f64())
}

impl FocusManager {
    /// New manager with default clock, nudge, and catch-up settings.
    #[must_use]
    pub fn new(
        familiar_id: impl Into<String>,
        store: Arc<dyn FocusStore>,
        subscriptions: Arc<dyn SubscriptionView>,
    ) -> Self {
        Self {
            familiar_id: familiar_id.into(),
            store,
            subscriptions,
            clock: monotonic_clock(),
            unread_nudge_enabled: true,
            nudge_debounce_seconds: DEFAULT_NUDGE_DEBOUNCE_S,
            catch_up_limit: DEFAULT_CATCH_UP_LIMIT,
            state: Mutex::new(FocusState {
                text_focus: None,
                voice_focus: None,
                channel_names: HashMap::new(),
                guild_names: HashMap::new(),
                last_nudge: f64::NEG_INFINITY,
            }),
            on_shift: Mutex::new(None),
        }
    }

    /// Inject a clock (default `Instant`-based monotonic seconds).
    #[must_use]
    pub fn with_clock(mut self, clock: Clock) -> Self {
        self.clock = clock;
        self
    }

    /// Enable / disable the unread nudge.
    #[must_use]
    pub const fn with_unread_nudge_enabled(mut self, enabled: bool) -> Self {
        self.unread_nudge_enabled = enabled;
        self
    }

    /// Set the nudge debounce window (seconds).
    #[must_use]
    pub const fn with_nudge_debounce_seconds(mut self, seconds: f64) -> Self {
        self.nudge_debounce_seconds = seconds;
        self
    }

    /// Set the catch-up window.
    #[must_use]
    pub const fn with_catch_up_limit(mut self, limit: usize) -> Self {
        self.catch_up_limit = limit;
        self
    }

    /// Load persisted focus pointers, dropping any no-longer-subscribed pointer.
    pub async fn initialize(&self) {
        let ptrs = self
            .store
            .get_focus_pointers(&self.familiar_id)
            .await
            .ok()
            .flatten();
        if let Some(ptrs) = ptrs {
            let text = self.keep_if_subscribed(ptrs.text_channel_id);
            let voice = self.keep_if_subscribed(ptrs.voice_channel_id);
            {
                let mut state = self.state.lock().expect("focus state mutex");
                state.text_focus = text;
                state.voice_focus = voice;
            }
            tracing::info!(
                "{} loaded {} {}",
                ls::tag("Focus", ls::LC),
                ls::kv_styled("text", &self.channel_label(text), ls::W, ls::LW),
                ls::kv_styled("voice", &self.channel_label(voice), ls::W, ls::LW),
            );
        }
    }

    fn keep_if_subscribed(&self, channel_id: Option<i64>) -> Option<i64> {
        match channel_id {
            None => None,
            Some(cid) if self.is_subscribed(cid) => Some(cid),
            Some(cid) => {
                tracing::warn!(
                    "{} dropped stale pointer {} (no longer subscribed)",
                    ls::tag("Focus", ls::LC),
                    ls::kv_styled("channel", &self.channel_label(Some(cid)), ls::W, ls::LW),
                );
                None
            }
        }
    }

    /// The staged-turn catch-up window (also the `shift_focus` preview size).
    #[must_use]
    pub const fn catch_up_limit(&self) -> usize {
        self.catch_up_limit
    }

    /// Current focus channel for `modality` (`"text"` → text pointer; any other
    /// string → voice pointer).
    #[must_use]
    pub fn get_focus(&self, modality: &str) -> Option<i64> {
        let state = self.state.lock().expect("focus state mutex");
        if modality == "text" {
            state.text_focus
        } else {
            state.voice_focus
        }
    }

    /// Whether `channel_id` is a known text/voice subscription.
    #[must_use]
    pub fn is_subscribed(&self, channel_id: i64) -> bool {
        u64::try_from(channel_id)
            .ok()
            .and_then(|cid| self.subscriptions.kind_for(cid))
            .is_some()
    }

    /// Sorted, deduped channel ids across text + voice subscriptions.
    #[must_use]
    pub fn subscribed_channels(&self) -> Vec<i64> {
        let mut set: BTreeSet<i64> = BTreeSet::new();
        for sub in self.subscriptions.all() {
            set.insert(i64::try_from(sub.channel_id).unwrap_or(i64::MAX));
        }
        set.into_iter().collect()
    }

    /// Whether `channel_id` is the active text or voice focus.
    #[must_use]
    pub fn is_focused(&self, channel_id: i64) -> bool {
        let state = self.state.lock().expect("focus state mutex");
        Some(channel_id) == state.text_focus || Some(channel_id) == state.voice_focus
    }

    /// Apply a focus shift immediately (at tool-call time).
    ///
    /// Modality is inferred from subscriptions — `voice` kind → voice; anything
    /// else (including unsubscribed) → text. A text shift promotes the target's
    /// staged backlog; both modalities move the pointer and persist both. The
    /// `on_shift` hook is awaited once, after the state lock is released.
    pub async fn shift_now(&self, channel_id: i64) {
        let kind = u64::try_from(channel_id)
            .ok()
            .and_then(|cid| self.subscriptions.kind_for(cid));
        let is_voice = kind == Some(SubscriptionKind::Voice);

        if is_voice {
            let (text, voice) = {
                let mut state = self.state.lock().expect("focus state mutex");
                state.voice_focus = Some(channel_id);
                (state.text_focus, state.voice_focus)
            };
            let _ = self
                .store
                .set_focus_pointers(&self.familiar_id, text, voice)
                .await;
            tracing::info!(
                "{} {}",
                ls::tag("\u{1f500} Focus", ls::LC),
                ls::kv_styled(
                    "voice",
                    &self.channel_label(Some(channel_id)),
                    ls::W,
                    ls::LW
                ),
            );
        } else {
            let promo = self
                .store
                .promote_staged_turns(&self.familiar_id, channel_id, self.catch_up_limit)
                .await
                .unwrap_or(Promotion {
                    consumed: 0,
                    missed: 0,
                });
            let (text, voice) = {
                let mut state = self.state.lock().expect("focus state mutex");
                state.text_focus = Some(channel_id);
                (state.text_focus, state.voice_focus)
            };
            let _ = self
                .store
                .set_focus_pointers(&self.familiar_id, text, voice)
                .await;
            tracing::info!(
                "{} {} {} {}",
                ls::tag("\u{1f500} Focus", ls::LC),
                ls::kv_styled("text", &self.channel_label(Some(channel_id)), ls::W, ls::LW),
                ls::kv_styled("promoted", &promo.consumed.to_string(), ls::W, ls::LG),
                ls::kv_styled("missed", &promo.missed.to_string(), ls::W, ls::LY),
            );
        }

        let hook = self.on_shift.lock().expect("on_shift mutex").clone();
        if let Some(hook) = hook {
            hook().await;
        }
    }

    /// Whether a non-focused arrival warrants a nudge (behavior 51).
    #[must_use]
    pub fn should_wake(&self, channel_id: i64) -> bool {
        if !self.unread_nudge_enabled {
            return false;
        }
        let now = (self.clock)();
        let state = self.state.lock().expect("focus state mutex");
        let focused = Some(channel_id) == state.text_focus || Some(channel_id) == state.voice_focus;
        if focused {
            return false;
        }
        (now - state.last_nudge) >= self.nudge_debounce_seconds
    }

    /// Record a nudge timestamp to start the debounce window.
    pub fn mark_nudge_pending(&self) {
        let now = (self.clock)();
        self.state.lock().expect("focus state mutex").last_nudge = now;
    }

    /// Responder end-of-turn hook — intentionally a no-op (behavior 50).
    #[allow(
        clippy::unused_async,
        reason = "kept async so both responders can await it uniformly"
    )]
    pub async fn end_turn(&self) {}

    /// Format a channel id as `#name(id)` or `#id` (`"none"` for `None`).
    #[must_use]
    pub fn channel_label(&self, channel_id: Option<i64>) -> String {
        channel_id.map_or_else(
            || "none".to_owned(),
            |cid| {
                let state = self.state.lock().expect("focus state mutex");
                state
                    .channel_names
                    .get(&cid)
                    .map_or_else(|| format!("#{cid}"), |name| format!("#{name}({cid})"))
            },
        )
    }

    /// Server name for `channel_id`; `None` for `None` input or unknown channel.
    #[must_use]
    pub fn guild_name_for(&self, channel_id: Option<i64>) -> Option<String> {
        let cid = channel_id?;
        self.state
            .lock()
            .expect("focus state mutex")
            .guild_names
            .get(&cid)
            .cloned()
    }

    /// Guild name for the current text focus; `None` when unset or unknown.
    #[must_use]
    pub fn presence_guild(&self) -> Option<String> {
        let state = self.state.lock().expect("focus state mutex");
        let cid = state.text_focus?;
        state.guild_names.get(&cid).cloned()
    }

    /// `#channel-name` (or `#<id>`) for the current text focus; `None` when unset.
    #[must_use]
    pub fn presence_text(&self) -> Option<String> {
        let state = self.state.lock().expect("focus state mutex");
        let cid = state.text_focus?;
        let name = state
            .channel_names
            .get(&cid)
            .cloned()
            .unwrap_or_else(|| cid.to_string());
        drop(state);
        Some(format!("#{name}"))
    }

    /// Seed a focus pointer without deferral, promotion, or persistence (startup).
    pub fn set_focus_immediately(&self, channel_id: i64, modality: &str) {
        {
            let mut state = self.state.lock().expect("focus state mutex");
            if modality == "text" {
                state.text_focus = Some(channel_id);
            } else {
                state.voice_focus = Some(channel_id);
            }
        }
        tracing::info!(
            "{} default {}",
            ls::tag("Focus", ls::LC),
            ls::kv_styled(
                modality,
                &self.channel_label(Some(channel_id)),
                ls::W,
                ls::LW
            ),
        );
    }

    /// Snapshot of the channel→display-name map (for the final-reminder digest).
    #[must_use]
    pub fn channel_names(&self) -> HashMap<i64, String> {
        self.state
            .lock()
            .expect("focus state mutex")
            .channel_names
            .clone()
    }

    /// Snapshot of the channel→server-name map (for the final-reminder digest;
    /// a [`PRIVATE_MESSAGE_GUILD_NAME`] entry marks a DM).
    #[must_use]
    pub fn guild_names(&self) -> HashMap<i64, String> {
        self.state
            .lock()
            .expect("focus state mutex")
            .guild_names
            .clone()
    }

    /// Set a channel's display name (populated by the Discord shell on ready).
    pub fn set_channel_name(&self, channel_id: i64, name: impl Into<String>) {
        self.state
            .lock()
            .expect("focus state mutex")
            .channel_names
            .insert(channel_id, name.into());
    }

    /// Set a channel's guild name (populated by the Discord shell on ready).
    pub fn set_guild_name(&self, channel_id: i64, name: impl Into<String>) {
        self.state
            .lock()
            .expect("focus state mutex")
            .guild_names
            .insert(channel_id, name.into());
    }

    /// Install the presence-refresh hook awaited after each shift.
    pub fn set_on_shift(&self, hook: OnShift) {
        *self.on_shift.lock().expect("on_shift mutex") = Some(hook);
    }

    /// Whether an `on_shift` hook is installed.
    #[must_use]
    pub fn has_on_shift(&self) -> bool {
        self.on_shift.lock().expect("on_shift mutex").is_some()
    }
}
